### ISSUE TITLE
Inconsistent substitution behavior in CVC5 and Z3

### ISSUE THREAD

**Issue by thebesttv** [sosy-lab/java-smt#323]
It seems that the [`substitute`](https://github.com/sosy-lab/java-smt/blob/e748acee211bcb54bc8497e0761520e8a27fb59d/src/org/sosy_lab/java_smt/solvers/cvc5/CVC5FormulaManager.java#L141C32-L141C42) in `CVC5FormulaManager` runs until the formula does not change.
However, the [`substitute`](https://github.com/sosy-lab/java-smt/blob/e748acee211bcb54bc8497e0761520e8a27fb59d/src/org/sosy_lab/java_smt/solvers/z3/Z3FormulaManager.java#L203C43-L203C43) in `Z3FormulaManager` only calls `Native.substitute` once.

The former approach enters an endless loop when I try to change formula `(+ a 1)` by substituting `a` with `(+ a 1)`:
``` java
  SolverContext context = SolverContextFactory.createSolverContext(
          Configuration.defaultConfiguration(),
          LogManager.createNullLogManager(),
          ShutdownNotifier.createDummy(),
          SolverContextFactory.Solvers.Z3);

  FormulaManager fmgr = context.getFormulaManager();
  IntegerFormulaManager imgr = fmgr.getIntegerFormulaManager();

  NumeralFormula.IntegerFormula
          a = imgr.makeVariable("a"),
          one = imgr.makeNumber(1);
  Formula formula = imgr.add(a, one);
  System.out.println(formula);
  System.out.println(fmgr.substitute(formula, Map.of(a, formula)));
```
The program is interrupted with `SIGSEGV`.
``` text
(+ a 1)

Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```
When Z3 is used instead of CVC5, the program runs successfully:
``` text
(+ a 1)
(+ a 1 1)
```

I think it would be preferable to perform substitution only once in CVC5.

**Comment by thebesttv** [comment]
Also, it seems that the original substitution function in CVC5 (`io.github.cvc5.Term#substitute(io.github.cvc5.Term, io.github.cvc5.Term)`) does not ["stops after the first term is replaced"](https://github.com/sosy-lab/java-smt/blob/e748acee211bcb54bc8497e0761520e8a27fb59d/src/org/sosy_lab/java_smt/solvers/cvc5/CVC5FormulaManager.java#L152C13-L152C13):
``` java
        ...
        NumeralFormula.IntegerFormula
                a = imgr.makeVariable("a"),
                b = imgr.makeVariable("b");
        Formula formula = imgr.add(imgr.add(imgr.add(a, a), a), b);

        Term tFormula = getTerm(formula),
                tA = getTerm(a),
                tB = getTerm(b);
        System.out.println(tFormula);
        System.out.println(tFormula.substitute(tA, tB)); // original function in CVC5
    }

    private static Term getTerm(Formula formula) {
        Field fieldTerm = getField(formula.getClass(), "cvc5term");
        fieldTerm.setAccessible(true);
        try {
            Term term = (Term) fieldTerm.get(formula);
            return term;
        } catch (IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }
```
The output:
``` text
(+ (+ (+ a a) a) b)
(+ (+ (+ b b) b) b)
```

**Comment by kfriedberger** [comment]
Thanks for reporting. We will take a look.

**Comment by kfriedberger** [comment]
This issue should be fixed with 6fa4f92ef453001eda814cd67c501e5fce2e542a.
We will include this fix in the next release.
